File: android/core/src/main/kotlin/app/kidschedule/core/sync/Lww.kt
package app.kidschedule.core.sync

/** LWW 裁决输入:同一行的一个版本。时间为 epoch 毫秒。协议 §6 */
data class LwwVersion(val clientUpdatedAt: Long, val deviceId: String)

enum class LwwVerdict { A_WINS, B_WINS, EQUAL_KEEP_A }

object Lww {
    /** a = 现存版本,b = 新来版本 */
    fun decide(a: LwwVersion, b: LwwVersion): LwwVerdict = when {
        b.clientUpdatedAt > a.clientUpdatedAt -> LwwVerdict.B_WINS
        b.clientUpdatedAt < a.clientUpdatedAt -> LwwVerdict.A_WINS
        b.deviceId > a.deviceId -> LwwVerdict.B_WINS
        b.deviceId < a.deviceId -> LwwVerdict.A_WINS
        else -> LwwVerdict.EQUAL_KEEP_A
    }
}

File: ios/KidSchedule/Core/Lww.swift
import Foundation

/// LWW 裁决输入:同一行的一个版本。时间为 epoch 毫秒。协议 §6
struct LwwVersion {
    let clientUpdatedAt: Int64
    let deviceId: String
}

enum LwwVerdict {
    case aWins, bWins, equalKeepA
}

enum Lww {
    /// a = 现存版本,b = 新来版本
    static func decide(a: LwwVersion, b: LwwVersion) -> LwwVerdict {
        if b.clientUpdatedAt > a.clientUpdatedAt { return .bWins }
        if b.clientUpdatedAt < a.clientUpdatedAt { return .aWins }
        if b.deviceId > a.deviceId { return .bWins }
        if b.deviceId < a.deviceId { return .aWins }
        return .equalKeepA
    }
}
